[SPARK-4286] Add an external shuffle service that can be run as a daemon.

dragos opened the pull request:
This allows Mesos deployments to use the shuffle service (and implicitly dynamic allocation). It does so by adding a new "main" class and two corresponding scripts in `sbin`:
- `sbin/start-shuffle-service.sh`
- `sbin/stop-shuffle-service.sh`

Specific options can be passed in `SPARK_SHUFFLE_OPTS`.

This is picking up work from #3861 /cc @tnachen

dragos committed:
Add an external shuffle service that can be run as a daemon.

This allows Mesos deployments to use the shuffle service (and
dynamic allocation).

AmplabJenkins commented:
Can one of the admins verify this patch?

rxin commented:
Thanks for submitting this. Can you add [Mesos] to the title?

tnachen commented:
This LGTM

srowen commented:
Hm, since this is adding a bit of a new API / service, I don't feel qualified to merge this myself. I'd prefer to get a second nod from @rxin

rxin commented on core/src/main/scala/org/apache/spark/shuffle/ExternalShuffleService.scala:
```diff
@@ -0,0 +1,87 @@
+ *
+ * @note This adds handlers for SIGINT and SIGTERM to terminate gracefully.
+ */
+object ExternalShuffleService extends Logging {
```
can we just add a main function to StandaloneWorkerShuffleService?

tnachen replied:
I remember I was told earlier to add a new class and try not to reuse the standalone one.
It's definitely preferred if we can just add a main method and call that instead.

rxin replied:
Can you cc who provided that comment? Maybe there are some reasoning behind that I haven't thought of yet.

tnachen replied:
@pwendell @andrewor14 can you guys comment on this? I can't remember exactly why, but perhaps I've misunderstood it.

dragos replied:
I think the discussion on #3861 started around [here](https://github.com/apache/spark/pull/3861#issuecomment-74207099)

andrewor14 replied:
How about a compromise, of renaming `StandaloneWorkerShuffleService` to `ExternalShuffleService` (since it has no standalone-specific references anyway) and adding a main method there? See my comment below.

dragos commented:
@pwendell @andrewor14 can I do anything to help push this forward?

andrewor14 commented:
Hi @dragos apologies for slipping a little. I have been occupied with other things and I will look at this within the next few days.

andrewor14 commented on core/src/main/scala/org/apache/spark/shuffle/ExternalShuffleService.scala:
```diff
@@ -0,0 +1,87 @@
+
+  private def installSignalHandlers() = {
+    new SigIntHandler("INT")
+    new SigIntHandler("TERM")
```
This seems lower level than is needed. Could we just add a shutdown hook so we don't have to deal with the specific signals?

dragos replied:
Sure. I just followed what's done in other services, but I'm happy to switch to shutdown hooks.

andrewor14 commented on docs/job-scheduling.md:
```diff
@@ -101,6 +101,11 @@ then set `yarn.nodemanager.aux-services.spark_shuffle.class` to
 
+If you wish to use the external shuffle service in a Mesos deployment you need to launch the
+service on each Mesos slave that might run Spark jobs. A launcher script is provided in the `sbin/`
+directory: `bin/start-shuffle-service.sh`. You can later stop the service by running
```
sbin

andrewor14 commented on docs/job-scheduling.md:
```diff
@@ -101,6 +101,11 @@ then set `yarn.nodemanager.aux-services.spark_shuffle.class` to
+service on each Mesos slave that might run Spark jobs. A launcher script is provided in the `sbin/`
+directory: `bin/start-shuffle-service.sh`. You can later stop the service by running
+`sbin/stop-shuffle-service.sh`. Additional configuration options can be passed in `SPARK_SHUFFLE_OPTS`.
+
```
This seems to be the wrong patch to put this. As of the changes here dynamic allocation is not yet supported in Mesos yet, so it would make more sense to put this paragraph in your other patch. Otherwise if this patch goes in first then the docs are a little misleading, especially since in L72 we explicitly say that DA on Mesos is not supported yet.

dragos replied:
My thinking was that any options can be passed using `SPARK_SHUFFLE_OPTS`, including `spark.shuffle.service.port` or authentication. I followed what's done for other services/servers, like the history server.

andrewor14 replied:
Wait, actually my comment wasn't referring specifically to `SPARK_SHUFFLE_OPTS`. I was referring to the whole paragraph, which is currently under the Dynamic Allocation section even though DA for Mesos isn't supported yet.

andrewor14 commented on core/src/main/scala/org/apache/spark/shuffle/ExternalShuffleService.scala:
```diff
@@ -0,0 +1,87 @@
+
+  private val barrier = new CountDownLatch(1)
+
+  def main(args: Array[String]): Unit = {
```
This whole method kind of duplicates the code in `StandaloneWorkerShuffleService`. If you look at that class, you'll find that there's actually nothing specific to standalone mode there. So I would suggest that we simply rename that class to `ExternalShuffleService` and add a main method there that just instantiates an instance of the service and blocks until the process exits.

dragos replied:
I agree there's duplication, but I went for it based on feedback on #3861 and advice from @tnachen. Probably misunderstanding on our part.

andrewor14 commented:
@dragos The changes look fine. I left a suggestion on the placement of functionality that I believe will make this service more general and at the same time reduce the amount of duplicate code. In terms of documentation, I would leave it out of this patch since we still need to document the existence of the shuffle service in the first place! (That is, outside the context of dynamic allocation)

Another thing is that the scope of this patch is actually outside of Mesos, so I would just remove references to that in the title.

andrewor14 commented on conf/spark-env.sh.template:
```diff
@@ -39,6 +39,7 @@
 # - SPARK_WORKER_DIR, to set the working directory of worker processes
 # - SPARK_WORKER_OPTS, to set config properties only for the worker (e.g. "-Dx=y")
 # - SPARK_HISTORY_OPTS, to set config properties only for the history server (e.g. "-Dx=y")
+# - SPARK_SHUFFLE_OPTS, to set config properties only for the external shuffle server (e.g. "-Dx=y")
```
elsewhere we call this a service, not a server

dragos replied:
Good point! Thanks

andrewor14 commented:
@dragos The changes look fine. I left a suggestion on the placement of functionality that I believe will make this service more general and at the same time reduce the amount of duplicate code. In terms of documentation, I would leave it out of this patch since we still need to document the existence of the shuffle service in the first place! (That is, outside the context of dynamic allocation)

Another thing is that the scope of this patch is actually outside of Mesos, so I would just remove references to that in the title.

dragos committed:
Reviewer’s comments:

- remove additional external shuffle class
- added companion object and main method to the standalone shuffle class
- change to a shutdown hook instead of signal handlers
- various typos

dragos commented:
I believe I addressed all comments except the one regarding `SPARK_SHUFFLE_OPTS`. I can remove the documentation for that feature. Should I keep the environment variable or should I remove that as well?

andrewor14 commented:
I just noticed we never ran tests for this. Jenkins test this please.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/ExternalShuffleService.scala:
```diff
@@ -18,12 +18,15 @@
 package org.apache.spark.deploy.worker
```
This is in the wrong package. It's no longer used for standalone mode only. It might make sense to put it in `o.a.s.shuffle`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/ExternalShuffleService.scala:
```diff
@@ -18,12 +18,15 @@
 import org.apache.spark.network.server.TransportServer
 import org.apache.spark.network.shuffle.ExternalShuffleBlockHandler
 
+import java.util.concurrent.CountDownLatch
```
this needs to be up there above the spark imports (see https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide)

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/ExternalShuffleService.scala:
```diff
@@ -32,7 +35,7 @@ import org.apache.spark.network.shuffle.ExternalShuffleBlockHandler
  * Optionally requires SASL authentication in order to read. See [[SecurityManager]].
  */
 private[worker]
```
`private[deploy]`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/ExternalShuffleService.scala:
```diff
@@ -64,3 +67,39 @@ class StandaloneWorkerShuffleService(sparkConf: SparkConf, securityManager: Secu
+    barrier.await()
+  }
+
+  private def installShutdownHook() = {
```
Please add `Unit` return type

SparkQA commented:
[Test build #30425 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30425/consoleFull) for   PR 4990 at commit [`4dc1f91`](https://github.com/apache/spark/commit/4dc1f9163b24b10d7ea71220384e556408ff657e).

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/ExternalShuffleService.scala:
```diff
@@ -64,3 +67,39 @@ class StandaloneWorkerShuffleService(sparkConf: SparkConf, securityManager: Secu
+
+    // we override this value since this service is started from the command line
+    // and we assume the user really wants it to be running
+    sparkConf.set("spark.shuffle.service.enabled", "true")
```
This is kind of weird. I would just add a `start` method in the shuffle service that ignores this flag. Really I think the flag should be checked outside of the shuffle service before instantiating an instance of it.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/ExternalShuffleService.scala:
```diff
@@ -64,3 +67,39 @@ class StandaloneWorkerShuffleService(sparkConf: SparkConf, securityManager: Secu
+    // we override this value since this service is started from the command line
+    // and we assume the user really wants it to be running
+    sparkConf.set("spark.shuffle.service.enabled", "true")
+    server = new ExternalShuffleService(sparkConf, securityManager)
```
we never started this service!

andrewor14 commented:
@dragos Looking good. I believe this will be the last round of reviews before we can merge this.

SparkQA commented:
[Test build #30425 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30425/consoleFull) for   PR 4990 at commit [`4dc1f91`](https://github.com/apache/spark/commit/4dc1f9163b24b10d7ea71220384e556408ff657e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30425/
Test PASSed.

dragos commented:
Great! I'm traveling today and tomorrow, so I'll address the remaining issues until early next week. Thanks!

> On 16 apr. 2015, at 19:40, andrewor14 <notifications@github.com> wrote:
> 
> @dragos Looking good. I believe this will be the last round of reviews before we can merge this.
> 
> —
> Reply to this email directly or view it on GitHub.
>

dragos committed:
Moved ExternalShuffleService to the `deploy` package + other minor tweaks.

dragos commented:
Let me know if there's anything else!

andrewor14 commented:
Jenkins test this please

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala:
```diff
@@ -15,14 +15,17 @@
 import org.apache.spark.network.server.TransportServer
 import org.apache.spark.network.shuffle.ExternalShuffleBlockHandler
+import java.util.concurrent.CountDownLatch
+import org.apache.spark.annotation.DeveloperApi
```
import order. I can fix this myself on merge

dragos replied:
Hm, I could swear I actually fixed this, but then again..

SparkQA commented:
[Test build #30600 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30600/consoleFull) for   PR 4990 at commit [`07804ad`](https://github.com/apache/spark/commit/07804adebcc9dd94723e2cf2df8de2a025d97308).

andrewor14 commented on launcher/src/main/java/org/apache/spark/launcher/SparkClassCommandBuilder.java:
```diff
@@ -69,6 +69,10 @@
     } else if (className.equals("org.apache.spark.executor.MesosExecutorBackend")) {
       javaOptsKeys.add("SPARK_EXECUTOR_OPTS");
       memKey = "SPARK_EXECUTOR_MEMORY";
+    } else if (className.equals("org.apache.spark.shuffle.ExternalShuffleService")) {
```
this is not the correct path, right?

dragos replied:
No its not. I'll take care of it.

andrewor14 commented:
LGTM, I will fix up the two comments I left when I merge this once tests pass.

dragos committed:
Import order and wrong name fixup.

SparkQA commented:
[Test build #30600 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30600/consoleFull) for   PR 4990 at commit [`07804ad`](https://github.com/apache/spark/commit/07804adebcc9dd94723e2cf2df8de2a025d97308).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30600/
Test PASSed.

dragos commented:
@andrewor14 I hope my latest changes are good to go. Let me know if there is anything else.

AmplabJenkins commented:
Can one of the admins verify this patch?

andrewor14 commented:
@dragos sorry for slipping on this. I've been busy with my own patches for a while. I just merged this into master. Thanks for your patience.

dragos commented:
No worries, thanks for reviewing it!
